plugins/solar/fronius

```diff
@@ -0,0 +1,285 @@
+#            "rel_SelfConsumption" : null
+#         },
+#         "Version" : "12"
+#      }
+#   },
+#   "Head" : {
+#      "RequestArguments" : {},
+#      "Status" : {
+#         "Code" : 0,
+#         "Reason" : "",
+#         "UserMessage" : ""
+#      },
+#      "Timestamp" : "2020-06-11T10:55:21+10:00"
+#   }
+#}
+#'
+#
+## http://fronius/solar_api/v1/GetActiveDeviceInfo.cgi?DeviceClass=SensorCard
+#GetActiveDeviceInfo='
+#{
+#   "Body" : {
+#      "Data" : {}
+#   },
+#   "Head" : {
+#      "RequestArguments" : {
+#         "DeviceClass" : "SensorCard"
+#      },
+#      "Status" : {
+#         "Code" : 0,
+#         "Reason" : "",
+#         "UserMessage" : ""
+#      },
+#      "Timestamp" : "2020-06-11T10:55:24+10:00"
+#   }
+#}
+#'
+#
+## http://fronius/solar_api/v1/GetLoggerConnectionInfo.cgi
+#GetLoggerConnectionInfo='
+#{
+#   "Body" : {
+#      "Data" : {
+#         "SolarNetConnectionState" : 2,
+#         "SolarWebConnectionState" : 2,
+#         "WLANConnectionState" : 2
+#      }
+#   },
+#   "Head" : {
+#      "RequestArguments" : {},
+#      "Status" : {
+#         "Code" : 0,
+#         "Reason" : "",
+#         "UserMessage" : ""
+#      },
+#      "Timestamp" : "2020-06-11T10:55:25+10:00"
+#   }
+#}
+#'
+
+# shellcheck disable=SC1090
+. "${MUNIN_LIBDIR}/plugins/plugin.sh"
+
+if [ "${MUNIN_DEBUG:-0}" = 1 ]; then
+    set -x
+fi
+
+INVERTER_BASE_URL=${inverter_base_url:-http://fronius}
+HOST_NAME=${host_name:-}
+CONNECT_TIMEOUT=${connect_timeout:-1}
+
+check_deps() {
+	for CMD in curl jq recode; do
+		if ! command -v "${CMD}" >/dev/null; then
+			echo "no (${CMD} not found)"
+		fi
+	done
+}
+
+CURL_ARGS="-s --connect-timeout ${CONNECT_TIMEOUT}"
+fetch() {
+	# shellcheck disable=SC2086
+	curl -f ${CURL_ARGS} "$@" \
+		|| { echo "error fetching ${*}" >&2; false; }
+}
+
+get_inverter_info() {
+	fetch "${INVERTER_BASE_URL}/solar_api/v1/GetInverterInfo.cgi" \
+	| recode html..ascii
+}
+
+get_power_flow_realtime_data() {
+	fetch "${INVERTER_BASE_URL}/solar_api/v1/GetPowerFlowRealtimeData.fcgi"
+	#echo "${GetPowerFlowRealtimeData}
+}
+
+# Run the command and arguments passed as arguments to this method, and cache
+# the response.  The first argument is a timeout (in minutes) after which the
+# cache is ignored and a new request is attempted.  If the request fails, the
+# cache is used. If the timeout is 0, the request is always attempted, using
+# the cache as a backup on failure.
+cached() {
+	timeout="${1}"
+	shift
+	fn="${1}"
+	shift
+	args=${*:-}
+
+	# shellcheck disable=SC2039
+	api_data=''
+	# shellcheck disable=SC2039
+	cachefile="${MUNIN_PLUGSTATE}/$(basename "${0}").${fn}.cache.json"
+	if [ -n "$(find "${cachefile}" -mmin "-${timeout}" 2>/dev/null)" ]; then
+		api_data=$(cat "${cachefile}")
+	else
+		# shellcheck disable=SC2086
+		api_data="$("${fn}" ${args} \
+			|| true)"
+
+		if [ -n "${api_data}" ]; then
+			echo "${api_data}" > "${cachefile}"
+		else
+			api_data=$(cat "${cachefile}")
```

Comment:
This may fail, if the cache file was never created, or?

Reply:
Correct, but there's not much we can do then. Do you have a suggestion about a better way to handle this? At the moment, the plugin just dies until it has obtained data at least once.